File: numass-core/src/main/java/inr/numass/NumassEnvelopeType.java
package inr.numass;

import hep.dataforge.io.envelopes.*;
import hep.dataforge.values.Value;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.nio.ByteBuffer;
import java.util.HashMap;
import java.util.Map;

import static hep.dataforge.io.envelopes.DefaultEnvelopeType.DEFAULT_ENVELOPE_TYPE;

/**
 * An envelope type for legacy numass tags. Reads legacy tag and writes DF02 tags
 */
public class NumassEnvelopeType implements EnvelopeType {

    public static final byte[] LEGACY_START_SEQUENCE = {'#', '!'};
    public static final byte[] LEGACY_END_SEQUENCE = {'!', '#', '\r', '\n'};

    @Override
    public int getCode() {
        return DEFAULT_ENVELOPE_TYPE;
    }

    @Override
    public String getName() {
        return "numass";
    }

    @Override
    public String description() {
        return "Numass legacy envelope";
    }

    @Override
    public EnvelopeReader getReader(Map<String, String> properties) {
        return new NumassEnvelopeReader();
    }

    @Override
    public EnvelopeWriter getWriter(Map<String, String> properties) {
        return new DefaultEnvelopeWriter(this, MetaType.resolve(properties));
    }

    public static class LegacyTag extends EnvelopeTag {
        @Override
        protected byte[] getStartSequence() {
            return LEGACY_START_SEQUENCE;
        }

        @Override
        protected byte[] getEndSequence() {
            return LEGACY_END_SEQUENCE;
        }

        /**
         * Get the length of tag in bytes. -1 means undefined size in case tag was modified
         *
         * @return
         */
        public int getLength() {
            return 30;
        }

        /**
         * Read leagscy version 1 tag without leading tag head
         *
         * @param buffer
         * @return
         * @throws IOException
         */
        protected Map<String, Value> readHeader(ByteBuffer buffer) throws IOException {
            Map<String, Value> res = new HashMap<>();

            int type = buffer.getInt(2);
            res.put(Envelope.TYPE_KEY, Value.of(type));

            short metaTypeCode = buffer.getShort(10);
            MetaType metaType = MetaType.resolve(metaTypeCode);

            if (metaType != null) {
                res.put(Envelope.META_TYPE_KEY, Value.of(metaType.getName()));
            } else {
                LoggerFactory.getLogger(EnvelopeTag.class).warn("Could not resolve meta type. Using default");
            }

            long metaLength = Integer.toUnsignedLong(buffer.getInt(14));
            res.put(Envelope.META_LENGTH_KEY, Value.of(metaLength));
            long dataLength = Integer.toUnsignedLong(buffer.getInt(22));
            res.put(Envelope.DATA_LENGTH_KEY, Value.of(dataLength));
            return res;
        }
    }

    private static class NumassEnvelopeReader extends DefaultEnvelopeReader {
        @Override
        protected EnvelopeTag newTag() {
            return new LegacyTag();
        }
    }
}

File: numass-core/src/main/kotlin/inr/numass/NumassEnvelopeType.kt
package inr.numass

import hep.dataforge.io.envelopes.*
import hep.dataforge.values.Value
import hep.dataforge.values.parseValue
import org.slf4j.LoggerFactory
import java.io.IOException
import java.nio.ByteBuffer
import java.nio.channels.FileChannel
import java.nio.file.Path
import java.nio.file.StandardOpenOption
import java.util.*

/**
 * An envelope type for legacy numass tags. Reads legacy tag and writes DF02 tags
 */
class NumassEnvelopeType : EnvelopeType {

    override val code: Int = DefaultEnvelopeType.DEFAULT_ENVELOPE_CODE

    override val name: String = "numass"

    override fun description(): String = "Numass legacy envelope"

    /**
     * Read as legacy
     */
    override fun getReader(properties: Map<String, String>): EnvelopeReader {
        return NumassEnvelopeReader()
    }

    /**
     * Write as default
     */
    override fun getWriter(properties: Map<String, String>): EnvelopeWriter {
        return DefaultEnvelopeWriter(this, MetaType.resolve(properties))
    }

    class LegacyTag : EnvelopeTag() {
        override val startSequence: ByteArray
            get() = LEGACY_START_SEQUENCE

        override val endSequence: ByteArray
            get() = LEGACY_END_SEQUENCE

        /**
         * Get the length of tag in bytes. -1 means undefined size in case tag was modified
         *
         * @return
         */
        override val length: Int
            get() = 30

        /**
         * Read leagscy version 1 tag without leading tag head
         *
         * @param buffer
         * @return
         * @throws IOException
         */
        override fun readHeader(buffer: ByteBuffer): Map<String, Value> {
            val res = HashMap<String, Value>()

            val type = buffer.getInt(2)
            res[Envelope.TYPE_PROPERTY] = Value.of(type)

            val metaTypeCode = buffer.getShort(10)
            val metaType = MetaType.resolve(metaTypeCode)

            if (metaType != null) {
                res[Envelope.META_TYPE_PROPERTY] = metaType.name.parseValue()
            } else {
                LoggerFactory.getLogger(EnvelopeTag::class.java).warn("Could not resolve meta type. Using default")
            }

            val metaLength = Integer.toUnsignedLong(buffer.getInt(14))
            res[Envelope.META_LENGTH_PROPERTY] = Value.of(metaLength)
            val dataLength = Integer.toUnsignedLong(buffer.getInt(22))
            res[Envelope.DATA_LENGTH_PROPERTY] = Value.of(dataLength)
            return res
        }
    }

    private class NumassEnvelopeReader : DefaultEnvelopeReader() {
        override fun newTag(): EnvelopeTag {
            return LegacyTag()
        }
    }

    companion object {
        val INSTANCE = NumassEnvelopeType()

        val LEGACY_START_SEQUENCE = byteArrayOf('#'.toByte(), '!'.toByte())
        val LEGACY_END_SEQUENCE = byteArrayOf('!'.toByte(), '#'.toByte(), '\r'.toByte(), '\n'.toByte())

        /**
         * Replacement for standard type infer to include legacy type
         *
         * @param path
         * @return
         */
        fun infer(path: Path): EnvelopeType? {
            return try {
                FileChannel.open(path, StandardOpenOption.READ).use {
                    val buffer = it.map(FileChannel.MapMode.READ_ONLY, 0, 6)
                    val array = ByteArray(6)
                    buffer.get(array)
                    val header = String(array)
                    when {
                        //TODO use templates from appropriate types
                        header.startsWith("#!") -> NumassEnvelopeType.INSTANCE
                        header.startsWith("#~DFTL") -> TaglessEnvelopeType.INSTANCE
                        header.startsWith("#~") -> DefaultEnvelopeType.INSTANCE
                        else -> null
                    }
                }
            } catch (ex: Exception) {
                LoggerFactory.getLogger(EnvelopeType::class.java).warn("Could not infer envelope type of file {} due to exception: {}", path, ex)
                null
            }

        }

    }

}
